Refinement of Output for Mequal and Automated Policy Bundle Documentatation Generation

mertbugrabicak opened the pull request:
First off sorry for the gigantic PR! 

The output we eventually wanted to get required doing some data processing before runtime, where we use OPA to generate objects containing the policy annotations in the source code, map them to a clean JSON, and then include them in the Mequal bundle for policies to make use of.

So the main changes:
- Script to generate annotation objects using OPA inspect -a
- Script to map these annotation object fields to one clean json file to show all metadata
- Create and inject this json into the bundle upon container build
- Add a "main" bundle to dynamically collect and show results for all bundles included in Mequal
- eval.rego to get evaluations of all the policies for a given input
- full.rego to get both the evaluations and a list of all policies ran on said input
- Add possibility to define severity, level, and any other custom fields in policies and rules via metadata annotations
- Edit policies to reflect/use the additional metadata fields
- Edit policies to remove the leveling at the directory level, since no longer will be done this way.
- Define external policy bundles included into Mequal using a json config file
- Script to download all external policy bundles to put into Mequal
- Script to generate AsciiDocs of the Mequal policies to be published in the documentation
- Github action to automatically generate and publish Mequal policy bundle documentation to be put on the Mequal website

mertbugrabicak committed:
docs: update documents to reflect pivot to opa from conforma

mertbugrabicak committed:
feat: add additional metadata to bundle manifest and package annotations

mertbugrabicak committed:
feat: add scripts to generate bundle metadata and docs from opa inspect annotations

mertbugrabicak committed:
feat: add main bundle in order to display cross bundle information

mertbugrabicak committed:
feat: run bundle metadata generation and inclusion into mequal upon image build

mertbugrabicak committed:
Merge branch 'project-ncl:main' into edit-docs

mertbugrabicak committed:
feat: added metadata json inside of the bundle on image build to be accessed by policies

mertbugrabicak committed:
feat: edit mapper to include custom fields, rule severities, policy and rule levels

mertbugrabicak committed:
feat: add main endpoint that controls output from all bundles at once

mertbugrabicak committed:
feat: add and move around policy information in policies

mertbugrabicak committed:
fix: add if keywords to make v1 opa compatible

mertbugrabicak committed:
feat: add full endpoint to return both bundle info and eval results

mertbugrabicak committed:
fix: change cli to point to full endpoint in image run

mertbugrabicak committed:
feat: include prodsec bundle in evaluation results

mertbugrabicak committed:
fix: adjust code comment to clarify

mertbugrabicak committed:
feat: restructure mequal policies to exclude level from directory

mertbugrabicak committed:
feat: dynamically fetch and show eval results from all bundles included in mequal

mertbugrabicak committed:
feat: update the documentation generate script for bundles to use latest json structure

mertbugrabicak committed:
feat: check if OPA binary defined as env before running script

mertbugrabicak committed:
docs: auto generate docs for policy bundles included

mertbugrabicak committed:
fix: make annotation script executable

mertbugrabicak committed:
chore: credit gemini for doc generate script

mertbugrabicak committed:
feat: get and download external policy bundles from configuration file

mertbugrabicak committed:
docs: fetch external policies for auto generating documentation

mertbugrabicak committed:
feat: make download external bundles script executable

jonathanchristison commented on container_files/Containerfile:
```diff
@@ -4,22 +4,24 @@ EXPOSE 8181
 
 USER root
 COPY ./policy/ ./policy/
+COPY ./hack/ ./hack/
```
I think of `hack` as scripts to support building and expect to run locally, these are ran as part of the container image build, we should perhaps re-purpose `local-opa-run.sh` to be ran at container build time, rather than maintain two seperate approaches to building the policy bundles.

mertbugrabicak replied:
Yes, you're absolutely right about this. I'll move the scripts that the image depends on to a `scripts` folder, and in `hack` I can create scripts to prep the development environment by running those scripts locally. It currently relies on building and running the image for development and testing, since inside the image it generates that metadata object to be used.

With something like dev-setup.sh,  we generate the metadata object and put it into where it would be in the directory, so we can then use OPA locally to do our development. This script can also include downloading the helper functions to be used, so we can remove them from the code at least and let it live in another repo. (currently each policy repo has them copied over as a temporary solution, which is not good at all)

Both can then be marked by gitignore so they don't end up making it into the source code.

mertbugrabicak replied:
Separated them into separate scripts folder

jonathanchristison commented on hack/generate_bundle_info.py:
```diff
@@ -0,0 +1,222 @@
+# Generated with assistance from a large language model trained by Google.
```
Add `#!/bin/env python` and set executable, I'm not sure about the current header, it doesn't give much information or attribution about what was used

mertbugrabicak replied:
Sure, I can do that

mertbugrabicak replied:
Done as `#!/usr/bin/env python3`

mertbugrabicak replied:
Forgot to address the LLM comment: Gemini by default isn't really attaching any info to the code it generates, and I just asked how to credit it. And it said to do this way since versioning may differ in the background if a specific model etc is specified. So I just kept it broad like this

jonathanchristison committed:
feat(policy versioning): add script to version policy

In this commit:

* Added Gemini generated script to use current git revision as .manifest
  version
* Call this script before building policy bundle in container

jonathanchristison committed:
feat(policy versioning): set manifest version for podman build

In this commit:

* Allow overrides to the manifest revision
* Update readme for building
* Update github ci to override manifest revision

jonathanchristison commented on policy/main/.manifest:
```diff
@@ -0,0 +1,4 @@
+{
+    "revision": "v1.0.0",
```
This doesn't need fixing just yet as the container image contains the git sha but I created a PR for a temporary solution to having this also set in the policy manifest as I think we will forget to update this and it may catch us out if we distribute policies by alternative methods -  https://github.com/mertbugrabicak/mequal/pull/2 but feel free to ignore

mertbugrabicak replied:
Oh yes, this was something I wanted to do but forgot! We can definitely inject it during image build.

For the mequal bundle versioning, we can use our git sha, and for the external policies we define in our config json, I can create a revision field where we can concretely define what sha we are fetching the external policy from, and then inject those shas to their respective manifests.

mertbugrabicak replied:
Will take in your changes and work from there

mertbugrabicak replied:
Merged your changes, will now expand it to include external policies

mertbugrabicak replied:
Will leave this for now, to think of a better solution if possible for the external policy revisions

jonathanchristison committed:
chore(policies): opa fmt

In this commit:

* `opa fmt policy --v1-compatible -w`

jonathanchristison commented on policy/main/eval.rego:
```diff
@@ -0,0 +1,102 @@
+import rego.v1
+
+# Main evaluation rule to generate the desired "evaluation" output object
+evaluation := {
```
`opa fmt` changes this but as far as I can see there are no functional or syntax changes, only formatting, given we want to use regal in future I don't think we need to apply the [changes](https://github.com/mertbugrabicak/mequal/pull/3) but feel free

mertbugrabicak replied:
I think I should be able to do this

mertbugrabicak replied:
Merged fine

mertbugrabicak committed:
feat: separate image scripts, create actual dev scripts in hack, remove helpers and get externally

mertbugrabicak committed:
feat: separate binary downloads into a separate script for use in multiple locations

mertbugrabicak committed:
fix: dev setup script will also download binaries

mertbugrabicak committed:
feat: multi stage build with ubi minimal final image to reduce size

mertbugrabicak commented:
- Moved the scripts used during image build to a ./scripts/ directory
- Added dev-setup.sh and dev-cleanup.sh to automatically get local dev environment ready.
- Moved helper functions to own directory where it is fetched from there now:
https://github.com/project-ncl/mequal-helpers
- .gitignore for dev related stuff, external bundles pulled, etc
- Added shebang to python script
- Optimized the container for Mequal to reduce to half its size

mertbugrabicak commented:
@jonathanchristison if you can have another look tomorrow, hopefully it should be mergeable now. it's it a lot better shape

mertbugrabicak commented:
There is also a TODO to adapt the local opa run script to our latest bundle structure, i might also work on that tomorrow, but that can also wait

jonathanchristison commented:
@mertbugrabicak looks good to me and some nice additions with local scripts and reduced runtime image size :+1:
